File: tools/sbm-tree/sbm_tree.py
#!/usr/bin/env python3
"""
sbm-tree: Enterprise Merkle tree builder for SignedByMe

Builds a Merkle tree from leaf commitments using SHA-256.
Outputs root.json (for API publish) and witnesses/*.json (per-user).

IMPORTANT: This tool uses SHA-256 with "merkle:" domain separator,
matching the Rust implementation in merkle_hash.rs and the Python
API in membership.py.

Usage:
    python sbm_tree.py build --client-id acme_corp --purpose allowlist \
        --commitments commitments.csv --output ./output

Commitments CSV format (one hex commitment per line, no header):
    0x1234...
    0x5678...
"""

import argparse
import json
import os
import sys
import hashlib
from pathlib import Path
from typing import List, Tuple
from dataclasses import dataclass
import time

# Constants matching WITNESS_SPEC.md
TREE_DEPTH = 20
HASH_ALG = "sha256-merkle"  # Changed from "poseidon"
WITNESS_VERSION = 1

# Domain separators matching Rust merkle_hash.rs
MERKLE_DOMAIN = b"merkle:"
LEAF_DOMAIN = b"leaf:"


def merkle_hash_pair(left: bytes, right: bytes) -> bytes:
    """
    Hash two children to create a parent node.
    
    Uses: SHA256("merkle:" || left || right)
    
    This matches the Rust implementation in merkle_hash.rs:
    ```rust
    pub fn hash_pair(left: &[u8; 32], right: &[u8; 32]) -> [u8; 32] {
        let mut hasher = Sha256::new();
        hasher.update(MERKLE_DOMAIN);  // b"merkle:"
        hasher.update(left);
        hasher.update(right);
        hasher.finalize().into()
    }
    ```
    """
    return hashlib.sha256(MERKLE_DOMAIN + left + right).digest()


def hash_leaf(value: bytes) -> bytes:
    """
    Hash a leaf value.
    
    Uses: SHA256("leaf:" || value)
    """
    return hashlib.sha256(LEAF_DOMAIN + value).digest()


@dataclass
class PathSibling:
    """A sibling in the Merkle path"""
    hash: bytes  # 32 bytes
    is_right: bool  # True if sibling is on the right


@dataclass  
class MerklePath:
    """Merkle path from leaf to root"""
    siblings: List[PathSibling]
    
    def compute_root(self, leaf: bytes) -> bytes:
        """Compute root from leaf using this path"""
        current = leaf
        for sibling in self.siblings:
            if sibling.is_right:
                # sibling on right: H(current, sibling)
                current = merkle_hash_pair(current, sibling.hash)
            else:
                # sibling on left: H(sibling, current)
                current = merkle_hash_pair(sibling.hash, current)
        return current


class MerkleTree:
    """SHA-256 Merkle tree with zero padding"""
    
    def __init__(self, leaves: List[bytes], depth: int = TREE_DEPTH):
        if not leaves:
            raise ValueError("Cannot build empty tree")
        
        self.depth = depth
        
        # Pad to 2^depth leaves
        target_size = 2 ** depth
        self.original_count = len(leaves)
        
        if len(leaves) > target_size:
            raise ValueError(f"Too many leaves: {len(leaves)} > {target_size}")
        
        # Zero padding (32 zero bytes)
        zero_leaf = bytes(32)
        padded = leaves + [zero_leaf] * (target_size - len(leaves))
        
        # Build layers bottom-up
        self.layers = [padded]
        
        while len(self.layers[-1]) > 1:
            prev = self.layers[-1]
            next_layer = []
            for i in range(0, len(prev), 2):
                h = merkle_hash_pair(prev[i], prev[i + 1])
                next_layer.append(h)
            self.layers.append(next_layer)
        
        self.root = self.layers[-1][0]
    
    def get_path(self, leaf_index: int) -> MerklePath:
        """Get Merkle path for a leaf (0-indexed)"""
        if leaf_index < 0 or leaf_index >= self.original_count:
            raise ValueError(f"Invalid leaf index: {leaf_index}")
        
        siblings = []
        idx = leaf_index
        
        for layer in self.layers[:-1]:  # All layers except root
            # Sibling index
            if idx % 2 == 0:
                sibling_idx = idx + 1
                is_right = True  # sibling is on the right
            else:
                sibling_idx = idx - 1
                is_right = False  # sibling is on the left
            
            siblings.append(PathSibling(
                hash=layer[sibling_idx],
                is_right=is_right
            ))
            
            idx //= 2
        
        return MerklePath(siblings)


def parse_hex_commitment(hex_str: str) -> bytes:
    """Parse a hex commitment string to bytes"""
    hex_str = hex_str.strip()
    if hex_str.startswith('0x') or hex_str.startswith('0X'):
        hex_str = hex_str[2:]
    # Pad to 64 hex chars (32 bytes)
    hex_str = hex_str.zfill(64)
    return bytes.fromhex(hex_str)


def load_commitments(filepath: str) -> List[bytes]:
    """Load hex commitments from CSV (one per line)"""
    commitments = []
    with open(filepath, 'r') as f:
        for line_num, line in enumerate(f, 1):
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            try:
                commitment = parse_hex_commitment(line)
                if len(commitment) != 32:
                    print(f"Warning: Line {line_num} commitment is not 32 bytes", file=sys.stderr)
                    continue
                commitments.append(commitment)
            except Exception as e:
                print(f"Warning: Invalid commitment on line {line_num}: {e}", file=sys.stderr)
    return commitments


def build_tree(args):
    """Build Merkle tree and output root + witnesses"""
    depth = args.depth
    
    if depth != TREE_DEPTH:
        print(f"WARNING: Using depth={depth} (production requires depth={TREE_DEPTH})", file=sys.stderr)
        print(f"         Roots with depth!={TREE_DEPTH} will be REJECTED by the API", file=sys.stderr)
    
    # Load commitments
    print(f"Loading commitments from {args.commitments}...")
    commitments = load_commitments(args.commitments)
    
    if not commitments:
        print("Error: No valid commitments found", file=sys.stderr)
        sys.exit(1)
    
    print(f"Loaded {len(commitments)} commitments")
    
    # Build tree
    print(f"Building depth-{depth} Merkle tree (SHA-256 with 'merkle:' domain)...")
    tree = MerkleTree(commitments, depth=depth)
    print(f"Root: 0x{tree.root.hex()}")
    
    # Create output directory
    output_dir = Path(args.output)
    output_dir.mkdir(parents=True, exist_ok=True)
    witnesses_dir = output_dir / "witnesses"
    witnesses_dir.mkdir(exist_ok=True)
    
    # Generate root.json
    now = int(time.time())
    root_id = f"{args.client_id}-{args.purpose}-{now}"
    
    purpose_map = {
        "none": 0,
        "allowlist": 1,
        "issuer_batch": 2,
        "revocation": 3
    }
    purpose_id = purpose_map.get(args.purpose, 0)
    
    root_json = {
        "root_id": root_id,
        "client_id": args.client_id,
        "purpose": args.purpose,
        "purpose_id": purpose_id,
        "root": "0x" + tree.root.hex(),
        "hash_alg": HASH_ALG,
        "depth": depth,
        "not_before": now,
        "expires_at": now + (365 * 86400),  # 1 year
        "description": f"{args.purpose} tree with {len(commitments)} members",
        "member_count": len(commitments)
    }
    
    root_path = output_dir / "root.json"
    with open(root_path, 'w') as f:
        json.dump(root_json, f, indent=2)
    print(f"Wrote {root_path}")
    
    # Generate witnesses
    print(f"Generating {len(commitments)} witnesses...")
    for i, commitment in enumerate(commitments):
        path = tree.get_path(i)
        
        # Verify path is correct
        computed_root = path.compute_root(commitment)
        if computed_root != tree.root:
            print(f"ERROR: Path verification failed for leaf {i}!", file=sys.stderr)
            sys.exit(1)
        
        witness = {
            "version": WITNESS_VERSION,
            "client_id": args.client_id,
            "root_id": root_id,
            "purpose_id": purpose_id,
            "hash_alg": HASH_ALG,
            "depth": depth,
            "not_before": root_json["not_before"],
            "expires_at": root_json["expires_at"],
            # Note: leaf_index intentionally omitted (privacy)
            "siblings": ["0x" + s.hash.hex() for s in path.siblings],
            "path_bits": [1 if s.is_right else 0 for s in path.siblings]
        }
        
        witness_path = witnesses_dir / f"witness_{i:06d}.json"
        with open(witness_path, 'w') as f:
            json.dump(witness, f, indent=2)
    
    print(f"Wrote {len(commitments)} witnesses to {witnesses_dir}/")
    
    # Summary
    print("\n=== Summary ===")
    print(f"Root ID: {root_id}")
    print(f"Root: 0x{tree.root.hex()}")
    print(f"Hash Algorithm: {HASH_ALG}")
    print(f"Members: {len(commitments)}")
    print(f"Depth: {depth}")
    print(f"\nTo publish root:")
    print(f"  curl -X POST https://api.signedby.me/v1/roots \\")
    print(f"    -H 'X-API-Key: YOUR_API_KEY' \\")
    print(f"    -H 'Content-Type: application/json' \\")
    print(f"    -d @{root_path}")


def verify_witness(args):
    """Verify a witness against a root"""
    with open(args.witness, 'r') as f:
        witness = json.load(f)
    
    commitment = parse_hex_commitment(args.commitment)
    
    # Reconstruct path
    siblings = []
    for i, (sib_hex, is_right) in enumerate(zip(witness["siblings"], witness["path_bits"])):
        sib_bytes = parse_hex_commitment(sib_hex)
        siblings.append(PathSibling(
            hash=sib_bytes,
            is_right=(is_right == 1)
        ))
    path = MerklePath(siblings)
    
    # Compute root
    computed_root = path.compute_root(commitment)
    
    print(f"Commitment: {args.commitment}")
    print(f"Hash Algorithm: {witness.get('hash_alg', 'sha256-merkle')}")
    print(f"Computed root: 0x{computed_root.hex()}")
    
    if args.root:
        expected = parse_hex_commitment(args.root)
        if computed_root == expected:
            print("✓ Verification PASSED")
        else:
            print(f"✗ Verification FAILED")
            print(f"  Expected: {args.root}")
            sys.exit(1)


def main():
    parser = argparse.ArgumentParser(
        description="sbm-tree: Enterprise Merkle tree builder for SignedByMe"
    )
    subparsers = parser.add_subparsers(dest="command", required=True)
    
    # Build command
    build_parser = subparsers.add_parser("build", help="Build Merkle tree from commitments")
    build_parser.add_argument("--client-id", required=True, help="Enterprise client ID")
    build_parser.add_argument("--purpose", required=True, 
                             choices=["none", "allowlist", "issuer_batch", "revocation"],
                             help="Tree purpose")
    build_parser.add_argument("--commitments", required=True, help="Path to commitments CSV")
    build_parser.add_argument("--output", required=True, help="Output directory")
    build_parser.add_argument("--depth", type=int, default=TREE_DEPTH,
                             help=f"Tree depth (default: {TREE_DEPTH}, use smaller for testing)")
    
    # Verify command
    verify_parser = subparsers.add_parser("verify", help="Verify a witness")
    verify_parser.add_argument("--witness", required=True, help="Path to witness JSON")
    verify_parser.add_argument("--commitment", required=True, help="Leaf commitment (hex)")
    verify_parser.add_argument("--root", help="Expected root (hex, optional)")
    
    args = parser.parse_args()
    
    if args.command == "build":
        build_tree(args)
    elif args.command == "verify":
        verify_witness(args)


if __name__ == "__main__":
    main()

File: native/btcdid_core/src/bin/sbm_tree.rs
//! sbm-tree: Enterprise Merkle tree builder for SignedByMe
//!
//! Builds depth-20 Poseidon Merkle trees from commitment files.
//! Outputs root.json + witnesses/*.json for enterprise membership proofs.
//!
//! Usage:
//!   sbm-tree build --client-id acme_corp --purpose issuer_batch \
//!       --commitments commitments.csv --output ./output
//!
//!   sbm-tree verify --witness witness.json --commitment 0x... --root 0x...

use std::fs::{self, File};
use std::io::{BufRead, BufReader, Write};
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};

use clap::{Parser, Subcommand};
use serde::{Deserialize, Serialize};

// Use btcdid_core's membership module
use btcdid_core::membership::poseidon::FieldElement;
use btcdid_core::membership::merkle::{MerkleTree, MerklePath, verify_merkle_path};

const TREE_DEPTH: usize = 20;
const HASH_ALG: &str = "poseidon";
const WITNESS_VERSION: u32 = 1;

#[derive(Parser)]
#[command(name = "sbm-tree")]
#[command(about = "Enterprise Merkle tree builder for SignedByMe")]
struct Cli {
    #[command(subcommand)]
    command: Commands,
}

#[derive(Subcommand)]
enum Commands {
    /// Build a Merkle tree from commitments
    Build {
        /// Enterprise client ID
        #[arg(long)]
        client_id: String,
        
        /// Tree purpose (issuer_batch, allowlist, revocation)
        #[arg(long)]
        purpose: String,
        
        /// Path to commitments file (one hex per line)
        #[arg(long)]
        commitments: PathBuf,
        
        /// Output directory
        #[arg(long)]
        output: PathBuf,
        
        /// Tree depth (default: 20)
        #[arg(long, default_value_t = TREE_DEPTH)]
        depth: usize,
    },
    
    /// Verify a witness against a commitment
    Verify {
        /// Path to witness JSON
        #[arg(long)]
        witness: PathBuf,
        
        /// Leaf commitment (hex)
        #[arg(long)]
        commitment: String,
        
        /// Expected root (hex, optional)
        #[arg(long)]
        root: Option<String>,
    },
}

#[derive(Serialize)]
struct RootJson {
    root_id: String,
    client_id: String,
    purpose: String,
    purpose_id: u8,
    root: String,
    hash_alg: String,
    depth: usize,
    not_before: u64,
    expires_at: u64,
    description: String,
    member_count: usize,
}

#[derive(Serialize, Deserialize)]
struct WitnessJson {
    version: u32,
    client_id: String,
    root_id: String,
    purpose_id: u8,
    hash_alg: String,
    depth: usize,
    not_before: u64,
    expires_at: u64,
    leaf_index: usize,
    siblings: Vec<String>,
    path_bits: Vec<u8>,
}

#[derive(Serialize)]
struct MappingEntry {
    leaf_index: usize,
    commitment: String,
    witness_file: String,
}

fn purpose_to_id(purpose: &str) -> u8 {
    match purpose {
        "none" => 0,
        "allowlist" => 1,
        "issuer_batch" => 2,
        "revocation" => 3,
        _ => 0,
    }
}

fn parse_hex_fe(s: &str) -> Result<FieldElement, String> {
    let s = s.trim();
    let hex = if s.starts_with("0x") || s.starts_with("0X") {
        &s[2..]
    } else {
        s
    };
    
    // Pad to 64 chars
    let padded = format!("{:0>64}", hex);
    let bytes = hex::decode(&padded).map_err(|e| format!("Invalid hex: {}", e))?;
    
    if bytes.len() != 32 {
        return Err(format!("Expected 32 bytes, got {}", bytes.len()));
    }
    
    let mut arr = [0u8; 32];
    arr.copy_from_slice(&bytes);
    Ok(FieldElement::from_bytes_be(&arr))
}

fn fe_to_hex(fe: &FieldElement) -> String {
    format!("0x{}", hex::encode(fe.to_bytes_be()))
}

fn load_commitments(path: &PathBuf) -> Result<Vec<FieldElement>, String> {
    let file = File::open(path).map_err(|e| format!("Cannot open file: {}", e))?;
    let reader = BufReader::new(file);
    
    let mut commitments = Vec::new();
    for (line_num, line) in reader.lines().enumerate() {
        let line = line.map_err(|e| format!("Read error: {}", e))?;
        let line = line.trim();
        
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        
        match parse_hex_fe(line) {
            Ok(fe) => commitments.push(fe),
            Err(e) => eprintln!("Warning: Line {}: {}", line_num + 1, e),
        }
    }
    
    Ok(commitments)
}

fn build_tree(
    client_id: String,
    purpose: String,
    commitments_path: PathBuf,
    output_path: PathBuf,
    depth: usize,
) -> Result<(), String> {
    if depth != TREE_DEPTH {
        eprintln!("WARNING: Using depth={} (production requires depth={})", depth, TREE_DEPTH);
        eprintln!("         Roots with depth!={} will be REJECTED by the API", TREE_DEPTH);
    }
    
    // Load commitments
    println!("Loading commitments from {:?}...", commitments_path);
    let commitments = load_commitments(&commitments_path)?;
    
    if commitments.is_empty() {
        return Err("No valid commitments found".into());
    }
    
    println!("Loaded {} commitments", commitments.len());
    
    // Check size limit
    let max_leaves = 1usize << depth;
    if commitments.len() > max_leaves {
        return Err(format!("Too many commitments: {} > {} (2^{})", 
            commitments.len(), max_leaves, depth));
    }
    
    // Build tree
    println!("Building depth-{} Merkle tree (padding with zeros)...", depth);
    println!("This may take a while for depth=20...");
    
    let tree = MerkleTree::with_depth(commitments.clone(), depth);
    println!("Root: {}", fe_to_hex(&tree.root));
    
    // Create output directories
    fs::create_dir_all(&output_path).map_err(|e| format!("Cannot create output dir: {}", e))?;
    let witnesses_dir = output_path.join("witnesses");
    fs::create_dir_all(&witnesses_dir).map_err(|e| format!("Cannot create witnesses dir: {}", e))?;
    
    // Generate root.json
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();
    
    let root_id = format!("{}-{}-{}", client_id, purpose, now);
    let purpose_id = purpose_to_id(&purpose);
    
    let root_json = RootJson {
        root_id: root_id.clone(),
        client_id: client_id.clone(),
        purpose: purpose.clone(),
        purpose_id,
        root: fe_to_hex(&tree.root),
        hash_alg: HASH_ALG.into(),
        depth,
        not_before: now,
        expires_at: now + 365 * 86400, // 1 year
        description: format!("{} tree with {} members", purpose, commitments.len()),
        member_count: commitments.len(),
    };
    
    let root_path = output_path.join("root.json");
    let root_file = File::create(&root_path).map_err(|e| format!("Cannot create root.json: {}", e))?;
    serde_json::to_writer_pretty(root_file, &root_json)
        .map_err(|e| format!("Cannot write root.json: {}", e))?;
    println!("Wrote {:?}", root_path);
    
    // Generate witnesses + mapping
    println!("Generating {} witnesses...", commitments.len());
    let mut mapping = Vec::new();
    
    for (i, commitment) in commitments.iter().enumerate() {
        let path = tree.get_path(i);
        
        // Verify path is correct
        if !verify_merkle_path(commitment, &path, &tree.root) {
            return Err(format!("Path verification failed for leaf {}!", i));
        }
        
        let witness = WitnessJson {
            version: WITNESS_VERSION,
            client_id: client_id.clone(),
            root_id: root_id.clone(),
            purpose_id,
            hash_alg: HASH_ALG.into(),
            depth,
            not_before: root_json.not_before,
            expires_at: root_json.expires_at,
            leaf_index: i,
            siblings: path.siblings.iter().map(|s| fe_to_hex(&s.hash)).collect(),
            path_bits: path.siblings.iter().map(|s| if s.is_right { 1 } else { 0 }).collect(),
        };
        
        let witness_filename = format!("witness_{:06}.json", i);
        let witness_path = witnesses_dir.join(&witness_filename);
        let witness_file = File::create(&witness_path)
            .map_err(|e| format!("Cannot create witness: {}", e))?;
        serde_json::to_writer_pretty(witness_file, &witness)
            .map_err(|e| format!("Cannot write witness: {}", e))?;
        
        mapping.push(MappingEntry {
            leaf_index: i,
            commitment: fe_to_hex(commitment),
            witness_file: witness_filename,
        });
        
        if (i + 1) % 1000 == 0 {
            println!("  Generated {}/{} witnesses", i + 1, commitments.len());
        }
    }
    
    println!("Wrote {} witnesses to {:?}", commitments.len(), witnesses_dir);
    
    // Write mapping file
    let mapping_path = output_path.join("mapping.json");
    let mapping_file = File::create(&mapping_path)
        .map_err(|e| format!("Cannot create mapping.json: {}", e))?;
    serde_json::to_writer_pretty(mapping_file, &mapping)
        .map_err(|e| format!("Cannot write mapping.json: {}", e))?;
    println!("Wrote {:?}", mapping_path);
    
    // Summary
    println!("\n=== Summary ===");
    println!("Root ID: {}", root_id);
    println!("Root: {}", fe_to_hex(&tree.root));
    println!("Members: {}", commitments.len());
    println!("Depth: {}", depth);
    println!("\nTo publish root:");
    println!("  curl -X POST https://api.signedby.me/v1/roots \\");
    println!("    -H 'X-API-Key: YOUR_API_KEY' \\");
    println!("    -H 'Content-Type: application/json' \\");
    println!("    -d @{:?}", root_path);
    
    Ok(())
}

fn verify_witness(
    witness_path: PathBuf,
    commitment_hex: String,
    expected_root: Option<String>,
) -> Result<(), String> {
    // Load witness
    let file = File::open(&witness_path)
        .map_err(|e| format!("Cannot open witness: {}", e))?;
    let witness: WitnessJson = serde_json::from_reader(file)
        .map_err(|e| format!("Cannot parse witness: {}", e))?;
    
    // Parse commitment
    let commitment = parse_hex_fe(&commitment_hex)?;
    
    // Reconstruct path
    let mut siblings = Vec::new();
    for (sib_hex, &is_right) in witness.siblings.iter().zip(witness.path_bits.iter()) {
        siblings.push(btcdid_core::membership::merkle::PathSibling {
            hash: parse_hex_fe(sib_hex)?,
            is_right: is_right == 1,
        });
    }
    let path = MerklePath { siblings };
    
    // Compute root
    let computed_root = path.compute_root(&commitment);
    
    println!("Commitment: {}", commitment_hex);
    println!("Computed root: {}", fe_to_hex(&computed_root));
    
    if let Some(expected) = expected_root {
        let expected_fe = parse_hex_fe(&expected)?;
        if computed_root == expected_fe {
            println!("✓ Verification PASSED");
            Ok(())
        } else {
            println!("✗ Verification FAILED");
            println!("  Expected: {}", expected);
            Err("Root mismatch".into())
        }
    } else {
        println!("(No expected root provided, showing computed value only)");
        Ok(())
    }
}

fn main() {
    let cli = Cli::parse();
    
    let result = match cli.command {
        Commands::Build { client_id, purpose, commitments, output, depth } => {
            build_tree(client_id, purpose, commitments, output, depth)
        }
        Commands::Verify { witness, commitment, root } => {
            verify_witness(witness, commitment, root)
        }
    };
    
    if let Err(e) = result {
        eprintln!("Error: {}", e);
        std::process::exit(1);
    }
}
